core/src/tauri/command.rs

```diff
@@ -358,6 +360,77 @@ pub async fn rebase(state: tauri::State<'_, State>) -> Result<(), TauriError> {
     Ok(())
 }
 
+#[tauri::command]
+pub async fn acquire_locks(
+    state: tauri::State<'_, State>,
+    files: Vec<ModifiedFile>,
+    fail_if_locked: bool,
+) -> Result<bool, TauriError> {
+    // get github username
+    let mut res = state
+        .client
+        .get(format!("{}/repo/gh/user", state.server_url))
+        .send()
+        .await?;
+
+    if res.status().is_client_error() {
+        let body = res.text().await?;
+        return Err(TauriError { message: body });
+    }
+
+    let username = match res.json::<UserInfoResponse>().await {
+        Ok(res) => res.username,
+        Err(err) => {
+            return Err(TauriError {
+                message: format!("Failed to get github username: {}", err),
+            });
+        }
+    };
+
+    // only lock files that are unlocked
+    let mut paths: Vec<String> = Vec::new();
+    let mut already_locked = false;
+    for file in files {
```

Comment:
We'll need to add one more failure case here, which is when the file has been modified upstream *by a different user*, we shouldn't let the user lock the file either. This will prevent cases where if the file has already been modified on latest `main`, the user won't be able to get their branch into a conflicted state.

The "by a different user" part is important too, because Quick Submit lets you submit files to `main`, but we wouldn't want to block you from submitting the file again because it technically doesn't conflict.

Reply:
acquire_locks doesn't let you take a lock from someone else, it only emits warnings that it was unable to lock a file because someone else already has a lock on it. It should skip re-locking files you already have checked out and skip locking files that are owned by different users. Is this the behavior you're talking about, or do you mean if any of the requested files are already locked by someone else, we shouldn't lock anything at all?

Reply:
So if you're not up-to-date on `main`, there is a chance the file you're trying to lock is NOT locked, but IS modified in between your current commit and `main`. You should not be able to lock the file until you've pulled latest, so trying to lock it here should warn or fail depending on the failure bool.